src/lib/form-field/form-field.ts

```diff
@@ -220,24 +227,40 @@ export class MatFormField implements AfterViewInit, AfterContentInit, AfterConte
     return ngControl && (ngControl as any)[prop];
   }
 
-  /** Whether the form field has a placeholder. */
   _hasPlaceholder() {
     return !!(this._control.placeholder || this._placeholderChild);
   }
 
+  _hasLabel() {
+    return !!this._labelChild;
+  }
+
+  _shouldLabelFloat() {
+    return this._canLabelFloat && (this._control.shouldLabelFloat ||
```

Comment:
This logic seems different than before, it didn't consider `_canPlaceholderFloat` before

Reply:
If I remember correctly, some of the logic was done by combining CSS classes. I simplified it to make it easier to follow.